src/core/ContentWindowController.cpp

```diff
@@ -198,34 +198,43 @@ void ContentWindowController::moveTo( const QPointF& position,
     default:
         return;
     }
-    constrainPosition( coordinates );
+    constrainPosition_( coordinates );
 
     contentWindow_.setCoordinates( coordinates );
 }
 
-void ContentWindowController::constrainSize( QSizeF& windowSize ) const
+void ContentWindowController::constrainSize_( QSizeF& windowSize ) const
 {
-    const QSizeF max_size = MAX_SIZE * displayGroup_.getCoordinates().size();
-    const QSizeF min_size = MIN_SIZE * displayGroup_.getCoordinates().size();
-
-    if ( windowSize.width() > max_size.width() ||
-         windowSize.height() > max_size.height( ))
-        windowSize.scale( max_size, Qt::KeepAspectRatio );
+    QSizeF maxSize = contentWindow_.getContent()->getDimensions();
+    if( maxSize.isEmpty( ))
+        maxSize = displayGroup_.getCoordinates().size();
+    maxSize *= ContentWindow::getMaxContentScale();
+    maxSize.rwidth() *= contentWindow_.getZoomRect().size().width();
+    maxSize.rheight() *= contentWindow_.getZoomRect().size().height();
+    if( windowSize.width() > maxSize.width() ||
+        windowSize.height() > maxSize.height( ))
+    {
+        windowSize = contentWindow_.getCoordinates().size();
```

Comment:
This introduces a subtle change in behaviour, which I am not sure is entirely correct.
Please check the following use case is: let's say that maxSize is decreased by large amount because zoomRect or MaxContentScale is modified. Then a user might not be able to scale down the window, because the new windowSize he or she can input will only have a small delta, which would always be above maxSize. Then windowSize will be reset to the current value instead of being scaled down to maxSize. What do you think?